### PR TITLE
docs: add a new section on content organization

### PR DESCRIPTION
**Pull request by bajtos**
Based on the discussion in #5550, I am updating the documentation for developers with the definition of "building-blocks", a list of packages considered as building blocks, and finally instructions on how to treat these packages in our docs, CLI templates and examples.

After this PR is approved, I'll created follow-up stories to ensure proper use of the following packages:

- packages/context - already done in #5625
- packages/openapi-v3

The following foundation-level packages are already not depended on directly:
- packages/express
- packages/http-server
- packages/metadata
- packages/repository-json-schema

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by bajtos** [comment]
Please note this is just small incremental step on a longer journey, I expect us to do more doc-related improvement in follow-up stories and pull requests, depending on how the discussions in #5113, #5550 and #5549 evolve.

File: docs/site/DEVELOPING.md [hunk 1]
```diff
@@ -286,6 +286,46 @@ LoopBack 4 documentation is hosted inside this monorepo in the
...
+[loopback.io](https://loopback.io).
+
+The rule of thumb: if a higher-level package is re-exporting all public APIs
+from a lower-level package (e.g. `@loopback/core` re-exports
+`@loopback/context`) or is using another package as a implementation building
+block (e.g. `@loopback/rest` uses `@loopback/express`), then the lower-level
+package should be moved to "building blocks" and our documentation should be
+updated to reference the higher-level "framework" instead.
+
+List of packages that are considered as building blocks:
```
Comment (hacksparrow) [inline docs/site/DEVELOPING.md:320]: And the rest are framework-level pacakges? If yes, we should mention that, else it feels like the list of framework-level pacakges is missing.

Comment (bajtos) [reply docs/site/DEVELOPING.md:320]: Done in 69dcaa6

There is one thing that bugs me a bit: we also have utility packages like `@loopback/build` or `@loopback/testlab` that are technically not part of the framework, but they are not building blocks either in the sense that they should be mentioned in our framework documentation.

I guess "framework-level" is not the best term to use. Can we come up with a more accurate name?

Comment (dhmlau) [reply docs/site/DEVELOPING.md:320]: What about the `extensions` and `examples` packages? Those seem more like an adds-on.

Comment (bajtos) [reply docs/site/DEVELOPING.md:320]: > What about the `extensions` and `examples` packages? Those seem more like an adds-on.

Sure :+1:

What I am trying to accomplish here is to distinguish between "building blocks" packages that should not be referred to in our docs, CLI templates and example apps; and everything else. 

I used the term "framework-level" for "everything else", which is clearly misleading based on your feedback.

So, what would be a better term?

Comment (hacksparrow) [reply docs/site/DEVELOPING.md:320]: > I guess "framework-level" is not the best term to use. Can we come up with a more accurate name?

They need not belong in the framework packages. They can be classified as helper/utility packages.

File: docs/site/DEVELOPING.md [hunk 2]
```diff
@@ -286,6 +286,48 @@ LoopBack 4 documentation is hosted inside this monorepo in the
...
+writing docs. While this is good for capturing the knowledge in a text form, it
+also makes the documentation difficult to follow by people new to LoopBack.
+Newcomers get quickly overwhelmed by the amount of concepts they need to
+understand and all the different packages they need to know about.
+
+To prevent that issue, we have categorized all monorepo packages into two
+groups:
+
+1. Building-blocks are lower-level packages that are typically not consumed by
+   LoopBack projects directly, instead there is a higher-level package exposing
```
Comment (raymondfeng) [inline docs/site/DEVELOPING.md:302]: `LoopBack projects` -> `LoopBack applications` or `LoopBack application projects`.

File: docs/site/DEVELOPING.md [hunk 3]
```diff
@@ -286,6 +286,48 @@ LoopBack 4 documentation is hosted inside this monorepo in the
...
 pull request.
 
+### Organization of content
+
+LoopBack 4 is a large codebase organized into dozens of packages. We,
+developers, are often tempted to include a lot of details and wider context when
+writing docs. While this is good for capturing the knowledge in a text form, it
+also makes the documentation difficult to follow by people new to LoopBack.
+Newcomers get quickly overwhelmed by the amount of concepts they need to
+understand and all the different packages they need to know about.
```
Comment (raymondfeng) [inline docs/site/DEVELOPING.md:296]: We should also mention the balance between `modular design` and `ease of use`.

- Module design leads to great flexibility and composability. As the number of modules grows, it also increases the complexity as developers need to know what to choose and how to use.
- We need to target high-level packages for application developers without dragging in lower-level details.

File: docs/site/DEVELOPING.md [hunk 4]
```diff
@@ -286,6 +286,48 @@ LoopBack 4 documentation is hosted inside this monorepo in the
...
+
+To prevent that issue, we have categorized all monorepo packages into two
+groups:
+
+1. Building-blocks are lower-level packages that are typically not consumed by
```
Comment (raymondfeng) [inline docs/site/DEVELOPING.md:301]: `not consumed` -> `not directly consumed` or `not explicitly consumed`?

Comment (raymondfeng) [reply docs/site/DEVELOPING.md:301]: I wonder if we can find a better name for `building blocks`.  Maybe `foundation-level` packages to be symmetric with `framework-level`?

Comment (raymondfeng) [reply docs/site/DEVELOPING.md:301]: Or `library-level`?

File: docs/site/DEVELOPING.md (hunk 3 above)
Comment (raymondfeng) [reply docs/site/DEVELOPING.md:296]: We may want to identify several personas of different interest/entry to LB here:

- Platform/extension developers
- Application developers

File: docs/site/DEVELOPING.md (hunk 4 above)
Comment (bajtos) [reply docs/site/DEVELOPING.md:301]: > `not consumed` -> `not directly consumed` or `not explicitly consumed`?

English is not my mother tongue, but I believe that my current version is grammatically correct while conveying the same meaning.

```
lower-level packages that are typically not consumed by LoopBack projects directly
```

Comment (bajtos) [reply docs/site/DEVELOPING.md:301]: > I wonder if we can find a better name for `building blocks`. Maybe `foundation-level` packages to be symmetric with `framework-level`?

I like "foundation-level" + "framework-level".

In https://github.com/strongloop/loopback-next/pull/5659#discussion_r435239706, we were discussing with @hacksparrow that "framework-level" may not be the best description. However, I think with building blocks renamed to "foundation-level", the term "framework-level" is ok.

File: docs/site/DEVELOPING.md (hunk 3 above)
Comment (bajtos) [reply docs/site/DEVELOPING.md:296]: > We may want to identify several personas of different interest/entry to LB here:
>
> - Platform/extension developers
> - Application developers

Yeah, I agree that at some point, we need to step back and reorganize our docs to cater for different personas. As I see it, our current docs is mostly aimed at application developers. In this pull request, I am proposing how to make the docs easier to read for app developers. We can discuss how to better address other personas in https://github.com/strongloop/loopback-next/issues/5113.

File: docs/site/DEVELOPING.md (hunk 2 above)
Comment (bajtos) [reply docs/site/DEVELOPING.md:302]: Well, I think this applies to most extension projects too, but let's not dwell on details. I am going to use "LoopBack applications" for now, we can improve this later as we get a better picture about the documentation for extension authors.

File: docs/site/DEVELOPING.md (hunk 1 above)
Comment (bajtos) [reply docs/site/DEVELOPING.md:320]: I am adding the following clarification to the docs:

```
#### Framework-level packages

All packages not listed above are considered as framework-level.

We consider utilities like `@loopback/testlab` and example projects like
`@loopback/todo` as framework-level too.```

**Comment by bajtos** [comment]
@hacksparrow @dhmlau @raymondfeng thank you for the suggestions. I rewrote most of the new content to incorporate your feedback, does the new version LGTY now?

**Comment by hacksparrow** [comment]
How do we handle "breaking changes" in docs organization? For example a framework-level package underwent some refactoring and ended up becoming a foundation-level package.

File: docs/site/DEVELOPING.md (hunk 1 above)
Comment (hacksparrow) [reply docs/site/DEVELOPING.md:336]: > We consider utilities like `@loopback/testlab` and example projects like
`@loopback/todo` as framework-level too.

Doesn't feel right. Anything that LoopBack developers would not (maybe even never) use in their project feels very odd to be classified under framework-level packages. We need a third category - helpers, utility, along that line.

**Comment by bajtos** [comment]
> How do we handle "breaking changes" in docs organization? For example a framework-level package underwent some refactoring and ended up becoming a foundation-level package.

At the moment, we have a single version of docs only ("the latest").

I think we will eventually need to introduce versioning (e.g. docs for `@loopback/rest` version from 2020 vs. version from 2021`), especially once we introduce LTS versions (see https://github.com/strongloop/loopback-next/issues/4398). Perhaps such versioning would help?

In the current state, where there is a single version of docs only: When a framework-level package becomes foundation-level, I am envisioning the following process:
- docs, CLI templates, example apps are updated to framework-level APIs instead of the package that was moved to foundation-level
- if & where appropriate, we may add additional sections to explain how to migrate from the "old" to the "new" style - see e.g. https://loopback.io/doc/en/lb4/DataSource-generator.html#legacy-json-based-configuration

Is there any specific concern you have, @hacksparrow?

**Comment by hacksparrow** [comment]
My concern was looking at outdated references. Versioning should help.

> docs, CLI templates, example apps are updated to use framework-level APIs instead of referring to he package that was moved to foundation-level
> if & where appropriate, we may add additional sections to explain how to migrate from the "old" to the "new" style - see e.g. https://loopback.io/doc/en/lb4/DataSource-generator.html#legacy-json-based-configuration

Sounds good for now.

File: docs/site/DEVELOPING.md (hunk 1 above)
Comment (bajtos) [reply docs/site/DEVELOPING.md:336]: I guess this is subjective.

Let's land my PR as the first step, and then you can open a follow-up PR to discuss how to improve this initial categorization.

Comment (hacksparrow) [reply docs/site/DEVELOPING.md:336]: Sounds good.

File: docs/site/DEVELOPING.md [hunk 5]
```diff
@@ -286,6 +286,69 @@ LoopBack 4 documentation is hosted inside this monorepo in the
...
+#### Foundation-level packages
+
+List of packages that are considered as building blocks:
+
+- [packages/context](https://github.com/strongloop/loopback-next/tree/master/packages/context)
+- [packages/express](https://github.com/strongloop/loopback-next/tree/master/packages/express)
+- [packages/http-server](https://github.com/strongloop/loopback-next/tree/master/packages/http-server)
+- [packages/metadata](https://github.com/strongloop/loopback-next/tree/master/packages/metadata)
+- [packages/openapi-v3](https://github.com/strongloop/loopback-next/tree/master/packages/openapi-v3)
+- [packages/repository-json-schema](https://github.com/strongloop/loopback-next/tree/master/packages/repository-json-schema)
```
Comment (raymondfeng) [inline docs/site/DEVELOPING.md:343]: This package depends on `@loopback/core` and `@loopback/repository`. It should be moved out.

File: docs/site/DEVELOPING.md [hunk 6]
```diff
@@ -286,6 +286,69 @@ LoopBack 4 documentation is hosted inside this monorepo in the
...
+
+#### Foundation-level packages
+
+List of packages that are considered as building blocks:
+
+- [packages/context](https://github.com/strongloop/loopback-next/tree/master/packages/context)
+- [packages/express](https://github.com/strongloop/loopback-next/tree/master/packages/express)
+- [packages/http-server](https://github.com/strongloop/loopback-next/tree/master/packages/http-server)
+- [packages/metadata](https://github.com/strongloop/loopback-next/tree/master/packages/metadata)
+- [packages/openapi-v3](https://github.com/strongloop/loopback-next/tree/master/packages/openapi-v3)
```
Comment (raymondfeng) [inline docs/site/DEVELOPING.md:342]: Not sure about this one either:
```
"dependencies": {
    "@loopback/core": "^2.7.1",
    "@loopback/repository-json-schema": "^2.4.3",
}
```

File: docs/site/DEVELOPING.md (hunk 5 above)
Comment (bajtos) [reply docs/site/DEVELOPING.md:343]: > This package depends on `@loopback/core` and `@loopback/repository`. It should be moved out.

Please note the criteria is not whether a package depends on something else from `@loopback/*` namespace, but whether LB application developers should depend on that package directly.

Both `@loopback/repository-json-schema` and `@loopback/openapi-v3` are not meant to be used directly. They are building blocks used by `@loopback/rest`. Applications are importing their APIs and functionality from `@loopback/rest`, as can be seen e.g. in [`examples/todo/src/controllers/todo.controller.ts`](https://github.com/strongloop/loopback-next/blob/master/examples/todo/src/controllers/todo.controller.ts)

**Pull request merged**